src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,323 @@
+            out ImmutableArray<MemberAccessExpressionSyntax> memberAccessExpressions,
+            CancellationToken cancellationToken)
+        {
+            tupleType = null;
+            memberAccessExpressions = default;
+
+            if (identifier.IsMissing)
+            {
+                return false;
+            }
+
+            if (!IsViableTupleTypeSyntax(typeNode))
+            {
+                return false;
+            }
+
+            var type = semanticModel.GetTypeInfo(typeNode, cancellationToken).Type;
+            if (type == null || !type.IsTupleType)
+            {
+                return false;
+            }
+
+            tupleType = (INamedTypeSymbol)type;
+            if (tupleType.TupleElements.Length < 2)
+            {
+                return false;
+            }
+
+            // All tuple elements must have been explicitly provided by the user.
+            foreach (var element in tupleType.TupleElements)
+            {
+                if (element.IsImplicitlyDeclared)
+                {
+                    return false;
+                }
+            }
+
+            var variableName = identifier.ValueText;
+
+            var references = ArrayBuilder<MemberAccessExpressionSyntax>.GetInstance();
+            try
+            {
+                // If the user actually uses the tuple local for anything other than accessing 
+                // fields off of it, then we can't deconstruct this tuple into locals.
+                if (!OnlyUsedToAccessTupleFields(
+                        semanticModel, searchScope, local, references, cancellationToken))
+                {
+                    return false;
+                }
+
+                // Can only deconstruct the tuple if the names we introduce won't collide
+                // with anything else in scope (either outside, or inside the method).
+                if (AnyTupleFieldNamesCollideWithExistingNames(
+                        semanticModel, tupleType, searchScope, cancellationToken))
+                {
+                    return false;
+                }
+
+                memberAccessExpressions = references.ToImmutable();
+                return true;
+            }
+            finally
+            {
+                references.Free();
+            }
+        }
+
+        private static bool AnyTupleFieldNamesCollideWithExistingNames(
+            SemanticModel semanticModel, INamedTypeSymbol tupleType,
+            SyntaxNode container, CancellationToken cancellationToken)
+        {
+            var existingSymbols = GetExistingSymbols(semanticModel, container, cancellationToken);
+
+            var reservedNames = semanticModel.LookupSymbols(container.SpanStart)
+                                             .Select(s => s.Name)
+                                             .Concat(existingSymbols.Select(s => s.Name))
+                                             .ToSet();
+
+            foreach (var element in tupleType.TupleElements)
```

Comment:
expanded for perf?

Reply:
yup.  but i don't feel strongly about it.